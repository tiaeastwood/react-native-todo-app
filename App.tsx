import { useEffect, useState } from "react";
import { StatusBar } from "expo-status-bar";
import {
	StyleSheet,
	View,
	FlatList,
	Alert,
	TouchableWithoutFeedback,
	Keyboard,
} from "react-native";
import Header from "./components/header";
import TodoItem from "./components/todoItem";
import AddTodo from "./components/addTodo";
import { ITodoItem } from "./types";
import AsyncStorage from "@react-native-async-storage/async-storage";

export default function App() {
	const [todos, setTodos] = useState<ITodoItem[]>([]);

	const pressHandler = (key: string) => {
		// removes the pressed todo from the list & returns new list
		setTodos((prevTodos) => {
			return prevTodos.filter((todo) => todo.key !== key);
		});
		removeTodo(key);
	};

	const storeTodo = async (value: ITodoItem) => {
		try {
			const jsonValue = JSON.stringify(value);
			await AsyncStorage.setItem(value.key, jsonValue);
		} catch (error) {
			console.log(error);
		}
	};

	const getTodo = async (key: string) => {
		try {
			const jsonValue = await AsyncStorage.getItem(key);
			return jsonValue != null ? JSON.parse(jsonValue) : null;
		} catch (error) {
			console.log(error);
		}
	};

	const getAllTodos = async () => {
		let keys: string[] = [];
		try {
			let foundKeys = await AsyncStorage.getAllKeys();
			keys = [...foundKeys];

			if (keys.length > 0) {
				let storedTodos: ITodoItem[] = [];

				for (const key of keys) {
					let foundTodo = await getTodo(key);
					storedTodos.push(foundTodo);
				}
				setTodos([...todos, ...storedTodos]);
			}
		} catch (error) {
			console.log(error);
		}
	};

	const removeTodo = async (key: string) => {
		try {
			await AsyncStorage.removeItem(key);
		} catch (error) {
			console.log(error);
		}
	};

	const submitHandler = (text: string) => {
		if (text.length > 3) {
			let newTodo = {
				text: text,
				key: Math.random().toString(),
			};
			setTodos((prevTodos) => {
				return [newTodo, ...prevTodos];
			});
			storeTodo(newTodo);
		} else {
			Alert.alert("Oops!", "Todos must be more than 3 characters long", [
				{ text: "OK", onPress: () => console.log("alert closed") },
			]);
		}
	};

	useEffect(() => {
		getAllTodos();
	}, []);

	return (
		<TouchableWithoutFeedback onPress={() => Keyboard.dismiss}>
			<View style={styles.container}>
				<StatusBar style="auto" />
				<Header />

				<View style={styles.content}>
					<AddTodo submitHandler={submitHandler} />
					{todos && (
						<View style={styles.list}>
							<FlatList
								data={todos}
								renderItem={({ item }) => (
									<TodoItem item={item} pressHandler={pressHandler} />
								)}
							/>
						</View>
					)}
				</View>
			</View>
		</TouchableWithoutFeedback>
	);
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		backgroundColor: "#fff",
		alignItems: "center",
		justifyContent: "center",
		width: "100%",
	},
	searchBar: {
		width: "50%",
		borderWidth: 1,
		padding: 10,
	},
	text: {
		fontWeight: "bold",
		fontSize: 50,
	},
	content: { padding: 40, width: "100%", flex: 1 },
	list: { marginTop: 20, flex: 1 },
});
