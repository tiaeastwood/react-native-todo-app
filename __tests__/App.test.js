import React from "react";
import { render, waitFor, screen } from "@testing-library/react-native";
import App from "../App.tsx";
import AsyncStorage from "@react-native-async-storage/async-storage";
import renderer from "react-test-renderer";

beforeEach(async () => {
	const value1 = { key: "1", text: "test" };
	const value2 = { key: "2", text: "another" };
	const value3 = { key: "3", text: "and another" };
	const jsonValue1 = JSON.stringify(value1);
	const jsonValue2 = JSON.stringify(value2);
	const jsonValue3 = JSON.stringify(value3);
	await AsyncStorage.setItem(value1.key, jsonValue1);
	await AsyncStorage.setItem(value2.key, jsonValue2);
	await AsyncStorage.setItem(value3.key, jsonValue3);
});

describe("<App />", () => {
	it("has 2 children", () => {
		const tree = renderer.create(<App />).toJSON();
		expect(tree.children.length).toBe(2);
	});

	it("renders correctly", () => {
		const tree = render(<App />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it("checks if Async Storage is used", async () => {
		render(<App />);

		waitFor(async () => {
			expect(AsyncStorage.getAllKeys).toHaveBeenCalled();
		});

		waitFor(async () => {
			expect(await AsyncStorage.getAllKeys()).toEqual(["1", "2", "3"]);
		});
	});
	it("renders ADD TODO button", async () => {
		render(<App />);
		expect(screen.getByText("ADD TODO")).toBeDefined();
	});
});
