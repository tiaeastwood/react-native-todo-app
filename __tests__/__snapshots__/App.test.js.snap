// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "coral",
        "display": "flex",
        "flexDirection": "row",
        "height": "auto",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingTop": 80,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 40,
        }
      }
    >
      My Todos
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 40,
        "width": "100%",
      }
    }
  >
    <View>
      <TextInput
        onChangeText={[Function]}
        placeholder="new todo..."
        style={
          Object {
            "borderBottomColor": "lightgrey",
            "borderBottomWidth": 1,
            "marginBottom": 10,
            "paddingHorizontal": 8,
            "paddingVertical": 6,
          }
        }
      />
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "coral",
            "borderRadius": 4,
            "elevation": 3,
            "justifyContent": "center",
            "paddingHorizontal": 32,
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": 0.25,
              "lineHeight": 21,
            }
          }
        >
          ADD TODO
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <RCTScrollView
        data={Array []}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        viewabilityConfigCallbackPairs={Array []}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</View>
`;
